### ISSUE TITLE
File.ReadAllBytes throws exception with a misleading message

### ISSUE THREAD

**Issue by alfredmyers** [dotnet/coreclr#18454]
# TL;DR;
The application calls `File.ReadAllBytes` passing a valid file name. For instance:

``` csharp
File.ReadAllBytes("93710.jpg");
```

Under some circumstances the application it ends up throwing a misleading exception.

> FileStream was asked to open a device that was not a file. For support for devices like 'com1:' or 'lpt1:', call CreateFile, then use the FileStream constructors that take an OS handle as an IntPtr.

# More Info
This is something I have found on the .NET Framework that I'm bringing here since there are talks about porting this application to .NET Core and the offending code is mostly the same.

Here's how the exception was logged on Windows:
```
Event code: 3005 
Event message: An unhandled exception has occurred. 
Event time: 6/6/2018 12:31:20 PM 
Event time (UTC): 6/6/2018 3:31:20 PM 
Event ID: 4c1127f18da24439abf85c192bb0ceeb 
Event sequence: 322 
Event occurrence: 1 
Event detail code: 0 
 
Application information: 
    Application domain: /LM/W3SVC/1/ROOT/Handlers-2-131727724516929646 
    Trust level: Full 
    Application Virtual Path: /Handlers 
    Application Path: F:\API\Handlers\ 
    Machine name: APISERVER-02 
 
Process information: 
    Process ID: 47224 
    Process name: w3wp.exe 
    Account name: domain\user 
 
Exception information: 
    Exception type: NotSupportedException 
    Exception message: FileStream was asked to open a device that was not a file. For support for devices like 'com1:' or 'lpt1:', call CreateFile, then use the FileStream constructors that take an OS handle as an IntPtr.
   at Microsoft.Win32.Win32Native.SafeCreateFile(String lpFileName, Int32 dwDesiredAccess, FileShare dwShareMode, SECURITY_ATTRIBUTES securityAttrs, FileMode dwCreationDisposition, Int32 dwFlagsAndAttributes, IntPtr hTemplateFile)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.File.InternalReadAllBytes(String path, Boolean checkHost)
   at Service.Handler.Web.Handlers.Image.<ProcessRequestAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.TaskAsyncHelper.EndTask(IAsyncResult ar)
   at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   at System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step)
   at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)

 
 
Request information: 
    Request URL: https://api.example.com:443/Handlers/Handler/Image.ashx?src=93710&length=563&fix=width 
    Request path: /Handlers/Handler/Image.ashx 
    User host address: 10.33.192.114 
    User:  
    Is authenticated: False 
    Authentication Type:  
    Thread account name: domain\user 
 
Thread information: 
    Thread ID: 145 
    Thread account name: domain\user 
    Is impersonating: False 
    Stack trace:    at Microsoft.Win32.Win32Native.SafeCreateFile(String lpFileName, Int32 dwDesiredAccess, FileShare dwShareMode, SECURITY_ATTRIBUTES securityAttrs, FileMode dwCreationDisposition, Int32 dwFlagsAndAttributes, IntPtr hTemplateFile)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.File.InternalReadAllBytes(String path, Boolean checkHost)
   at Service.Handler.Web.Handlers.Image.<ProcessRequestAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.TaskAsyncHelper.EndTask(IAsyncResult ar)
   at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   at System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step)
   at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
 
 ```

Here's the offending code on the .NET Framework (https://referencesource.microsoft.com/#mscorlib/microsoft/win32/win32native.cs):
``` csharp
// https://referencesource.microsoft.com/#mscorlib/microsoft/win32/win32native.cs,980

int fileType = Win32Native.GetFileType(handle);
if (fileType != Win32Native.FILE_TYPE_DISK) {
   handle.Dispose();
   throw new NotSupportedException(Environment.GetResourceString("NotSupported_FileStreamOnNonFiles"));
}

.
.
.

// https://referencesource.microsoft.com/#mscorlib/microsoft/win32/win32native.cs,1035

[DllImport(KERNEL32)]
[ResourceExposure(ResourceScope.None)]
internal static extern int GetFileType(SafeFileHandle handle);
```

And here's the equivalent code on .NET Core: 
https://github.com/dotnet/coreclr/blob/f4e09d1898a8eb172e5c286102ef9d1269b38666/src/System.Private.CoreLib/shared/System/IO/FileStream.Windows.cs#L57-L66

The thing is any error returned from [GetFileType](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364960.aspx) is ignored. In it's place, a NotSupportedException with a fixed (and possibly misleading) message is thrown making it harder to troubleshoot real cause of the exception.
